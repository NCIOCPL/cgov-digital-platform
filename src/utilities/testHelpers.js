import React from 'react';
import { MemoryRouter, Router } from 'react-router-dom';
import { mount } from 'enzyme';
import { createMemoryHistory } from 'history';
import { render } from '@testing-library/react';

export const TEST_ELEMENT_ID = 'test';

// renders an element in a test `document` and returns it.
export const getDomElement = element => {
  const instance = mount(<MemoryRouter>{element}</MemoryRouter>);
  const renderedElement = instance.getDOMNode();

  // Jest ships with `jsdom` which simulates a DOM environment. this environment
  // is cleared on every file run as each is its own thread... so, `document` will
  // reset per test file.
  document.getElementById(TEST_ELEMENT_ID).appendChild(renderedElement);

  return renderedElement;
};

// sets up a test DOM - a container to populate in tests.
export const setupDom = () => {
  document.body.innerHTML = `<div id="${TEST_ELEMENT_ID}"></div>`;
};

// restets a test DOM.
export const resetDom = () => {
  document.body.innerHTML = '';
};

export const renderWithRouter = (
  ui,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  } = {}
) => {
  return {
    ...render(<Router history={history}>{ui}</Router>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
};
