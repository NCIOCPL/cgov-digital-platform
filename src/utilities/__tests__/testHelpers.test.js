import React from 'react';
import { getDomElement, resetDom, setupDom } from '../testHelpers';
describe('testHelpers', () => {
  describe('getDomElement', () => {
    beforeEach(() => {
      setupDom();
    });
    afterEach(() => {
      resetDom();
    });
    it('returns a rendered element', () => {
      const element = getDomElement(<p>Ipsum lorem</p>);
      expect(element).toMatchSnapshot();
    });
    it('populates a rendered element in the DOM', () => {
      getDomElement(<p>Ipsum lorem</p>);
      expect(document.body).toMatchSnapshot();
    });
  });
  describe('setupDom', () => {
    it('renders a container element', () => {
      setupDom();
      expect(document.body).toMatchSnapshot();
    });
  });
  describe('resetDom', () => {
    it('removes container element from DOM', () => {
      resetDom();
      expect(document.body).toMatchSnapshot();
    });
  });
});
