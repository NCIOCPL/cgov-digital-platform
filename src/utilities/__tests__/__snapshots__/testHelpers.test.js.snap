// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testHelpers getDomElement populates a rendered element in the DOM 1`] = `
<body>
  <div
    id="test"
  >
    <p>
      Ipsum lorem
    </p>
  </div>
</body>
`;

exports[`testHelpers getDomElement returns a rendered element 1`] = `
<p>
  Ipsum lorem
</p>
`;

exports[`testHelpers resetDom removes container element from DOM 1`] = `<body />`;

exports[`testHelpers setupDom renders a container element 1`] = `
<body>
  <div
    id="test"
  />
</body>
`;
